src/compiler/checker.ts

```diff
@@ -45578,6 +45639,305 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+        // For a parameter of declared type `T` to be a valid reference for narrowing, it must satisfy:
+        // - the parameter name is an identifier
+        // - if the parameter is optional, then `T`'s constraint must allow for undefined
+        function getValidParameterReference(paramDecl: ParameterDeclaration, constraint: Type): Identifier | undefined {
+            if (!isIdentifier(paramDecl.name)) return;
+            const isOptional = !!paramDecl.questionToken || isJSDocOptionalParameter(paramDecl);
+            if (isOptional && !containsUndefinedType(constraint)) return;
+            return paramDecl.name;
+        }
+
+        function isReferenceToTypeParameter(typeParam: TypeParameter, node: TypeReferenceNode) {
+            return getTypeFromTypeReference(node) === typeParam;
+        }
+
+        function isTypeParameterReferenced(typeParam: TypeParameter, node: TypeNode) {
+            return isReferenced(node);
+
+            function isReferenced(node: Node): boolean {
+                if (isTypeReferenceNode(node)) {
+                    return isReferenceToTypeParameter(typeParam, node);
+                }
+                if (isTypeQueryNode(node)) {
+                    return isTypeParameterPossiblyReferenced(typeParam, node);
+                }
+                return !!forEachChild(node, isReferenced);
+            }
+        }
+    }
+
+    function isNarrowableReturnType(returnType: IndexedAccessType | ConditionalType): boolean {
+        return isConditionalType(returnType)
+            ? isNarrowableConditionalType(returnType)
+            : !!(returnType.indexType.flags & TypeFlags.TypeParameter);
+    }
+
+    function isNarrowableConditionalType(type: ConditionalType, mapper?: TypeMapper): boolean {
+        const typeArguments = mapper && map(type.root.outerTypeParameters, t => {
+            const mapped = getMappedType(t, mapper);
+            if (isNarrowingSubstitutionType(mapped)) {
+                return (mapped as SubstitutionType).baseType;
+            }
+            return mapped;
+        });
+        const id = `${type.id}:${getTypeListId(typeArguments)}`;
+        let result = narrowableReturnTypeCache.get(id);
+        if (result === undefined) {
+            const nonNarrowingMapper = type.root.outerTypeParameters
+                && typeArguments
+                && createTypeMapper(type.root.outerTypeParameters, typeArguments);
+            const instantiatedType = instantiateType(type, nonNarrowingMapper);
+            result = isConditionalType(instantiatedType) && isNarrowableConditionalTypeWorker(instantiatedType);
+            narrowableReturnTypeCache.set(id, result);
+        }
+        return result;
+    }
+
+    // A narrowable conditional type is one that has the following shape:
+    // `T extends A ? TrueBranch<T> : FalseBranch<T>`, in other words:
+    // (0) The conditional type is distributive;
+    // (1) The conditional type has no `infer` type parameters;

```

Comment:
Why's that? Is something like
```ts
function f<T extends [number, string] | [string, number]>(arg: T): T extends [number, infer U] ? U : never {
  if (typeof arg[0] === "number") {
    return arg[1];
  }
  throw new Error("Got non numeric tuple");
}
```
for some reason not feasible to check? Is it just because we'd have to do inference between the `true` branch type and the candidate return type to fill in the `infer` type variable, and that's not implemented yet? Because I could equivalently write
```ts
function f<T extends [number, U] | [string, number], U extends string>(arg: T): T extends [number, U] ? U : never {
  if (typeof arg[0] === "number") {
    return arg[1];
  }
  throw new Error("Got non numeric tuple");
}
```
which afaik is valid under these rules. I think all `infer` type parameters in types satisfying the rest of these rules can be rewritten like this pretty easily, which seems to imply the `infer` itself aughta work, but I could be wrong.

Reply:
Basically I don't know how inference would work safely between the substitution type I'm using to represent a narrowed type parameter and the infer type parameters, given how inference works today.

In `T extends [number, infer U] ? U : never`, if we instantiate `T` with substitution type `T & [number, string]`, we would infer from `T & [number, string]` to `[number, infer U]`.
Now, we can't simply infer `string` for `U`, because it would not be sound to narrow the return type to `string`:
 the caller can specify that `T` is `[number, "foo"]` and then `U` is inferred to be `"foo"`, not `string`.
 We'd have to infer something like `T[1] & string` or simply `T[1]` for `U` for this to be safe, and I think we don't have the ability to do that.